src/Shared/IInternable.cs

```diff
@@ -0,0 +1,346 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Text;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build
+{
+    #region IInternable
+    /// <summary>
+    /// Define the methods needed to intern something.
+    /// </summary>
+    internal interface IInternable
+    {
+        /// <summary>
+        /// The length of the target.
+        /// </summary>
+        int Length { get; }
+
+        /// <summary>
+        /// Indexer into the target. Presumed to be fast.
+        /// </summary>
+        char this[int index] { get; }
+
+        /// <summary>
+        /// Convert target to string. Presumed to be slow (and will be called just once).
+        /// </summary>
+        string ExpensiveConvertToString();
+
+        /// <summary>
+        /// Compare target to string. Assumes string is of equal or smaller length than target.
+        /// </summary>
+        bool StartsWithStringByOrdinalComparison(string other);
+
+        /// <summary>
+        /// Reference compare target to string. If target is non-string this should return false.
+        /// </summary>
+        bool ReferenceEquals(string other);
+    }
+    #endregion
+
+
+    #region IInternable Implementations
+    /// <summary>
+    /// A wrapper over StringBuilder.
+    /// </summary>
+    internal struct StringBuilderInternTarget : IInternable
+    {
+        /// <summary>
+        /// The held StringBuilder
+        /// </summary>
+        private readonly StringBuilder _target;
+
+        /// <summary>
+        /// Pointless comment about constructor.
+        /// </summary>
+        internal StringBuilderInternTarget(StringBuilder target)
+        {
+            _target = target;
+        }
+
+        /// <summary>
+        /// The length of the target.
+        /// </summary>
+        public int Length => _target.Length;
+
+        /// <summary>
+        /// Indexer into the target. Presumed to be fast.
```

Comment:
Warning--  indexing into a StringBuilder is NOT necessarily fast -- and looping through it (as one might through an array or string) can be n^2. (It depends on whether the StringBuilder is fragmented, which depends on how it was created and modified.) This is in the docs now, IIRC because MSBuild was doing this somewhere else.

Can indexing into a StringBuilder be avoided entirely?

Reply:
Here's context - https://github.com/Microsoft/msbuild/issues/1593

BTW, to compare to string -- on .NET Core there is a `StringBuilder.Equals(ReadOnlySpan<Char>)`. On .NET FRamework, you have to create a StringBuilder from the string, or vice versa, and compare like with like.

Reply:
Thank you. I'd like to keep the PR focused only on the interner itself. This code is just being copied verbatim from `OpportunisticIntern.cs`. I will address this separately later if you don't mind.

Reply:
Might be worth opening an issue if there isn't one already.

Reply:
This is now tracked with https://github.com/microsoft/msbuild/issues/5438.